### PR TITLE
Scaled dot-product attention

### PR DESCRIPTION
**Pull request by jindrahelcl**
Scaled dot-product attention from Vaswani et al. (2017): https://arxiv.org/pdf/1706.03762.pdf

This pull request also include the multi-head attention generalization.

First, #536 should be merged (not necessarily, but favorably)

### CODE REVIEW

**Comment by jlibovicky** [comment]
Můžeš to prosím rebasovat na mastera, ať vidím, co za změny patří jenom do tohodle PR?

**Comment by jindrahelcl** [comment]
ono to vychází z těch ensemblů. Ale je to tady nějaký podělaný

**Comment by jindrahelcl** [comment]
teď už to vypadá líp...

**Comment by jindrahelcl** [comment]
Btw. na mastera to rebasovat jde, ale vyžaduje to netriviální čas triviálních editačních prací (který by se pak znova musely řešit v konfliktech u #536)

File: neuralmonkey/attention/scaled_dot_product.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                      last_loop_state: MultiHeadLoopStateTA) -> None:
+        for i in range(self.n_heads):
+            head_weights = last_loop_state.head_weights[i].stack()
+            self.histories["{}_head{}".format(key, i)] = head_weights
+
+    @property
+    def context_vector_size(self) -> int:
+        return self.attention_values.get_shape()[-1].value
+
+    def visualize_attention(self, key: str) -> None:
```
Comment (jlibovicky) [inline neuralmonkey/attention/scaled_dot_product.py]: Jak se to liší od base attentionu?

File: neuralmonkey/attention/scaled_dot_product.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
+"""The scaled dot-product attention mechanism, defined in Vaswani et al. (2017)
```
Comment (jlibovicky) [inline neuralmonkey/attention/scaled_dot_product.py]: Myslím, že by si to zasloužilo delší docsting.

File: neuralmonkey/attention/scaled_dot_product.py (hunk 1 above)
Comment (jindrahelcl) [reply neuralmonkey/attention/scaled_dot_product.py]: je to celý o dimenzi větší. Za každou hlavu máš jednu historii. A to ještě přijde druhá dimenze za to, že v tom transformeru děláš attetnion pro několik queries současně.

File: neuralmonkey/attention/scaled_dot_product.py (hunk 2 above)
Comment (jindrahelcl) [reply neuralmonkey/attention/scaled_dot_product.py]: to je pravda

Comment (jlibovicky) [reply neuralmonkey/attention/scaled_dot_product.py]: Tak napiš dvěma větama, jak to funguje a už to zmerǧujeme.

File: neuralmonkey/attention/scaled_dot_product.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                                                          tf.Tensor]:
+        # shape: batch, time (similarities of attention keys in batch and time
+        # to the queries in the batch)
+        dot_product = tf.reduce_sum(
+            tf.expand_dims(query, 1) * keys, [-1])
+        energies = dot_product * self._scaling_factor
+
+        weights = tf.nn.softmax(energies)
+
+        if self.attention_mask is not None:
```
Comment (varisd) [inline neuralmonkey/attention/scaled_dot_product.py]: Nemel by se ten softmax aplikovat az po maskovani? (aspon tak to uvadeji v tom paperu)

File: neuralmonkey/attention/scaled_dot_product.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            weights_all = weights * self.attention_mask
+            norm = tf.reduce_sum(weights_all, 1, keep_dims=True) + 1e-8
+            weights = weights_all / norm
+
+        # apply dropout to the weights (Attention Dropout)
+        weights = dropout(weights, self.dropout_keep_prob, self.train_mode)
+
+        # sum up along the time axis
+        context = tf.reduce_sum(
+            tf.expand_dims(weights, -1) * values, [1])
```
Comment (varisd) [inline neuralmonkey/attention/scaled_dot_product.py]: Nebylo by lepsi tohle nahradit (podobne i na jinych mistech, kde delas dot product)?
context = tf.matmul(tf.expand_dims(weights, -1), values)

Nebo je tvoje reseni efektivnejsi?

File: neuralmonkey/attention/scaled_dot_product.py (hunk 3 above)
Comment (jlibovicky) [reply neuralmonkey/attention/scaled_dot_product.py]: One se to renormalizuje, takže je to skoro stejné jako kdyby bylo před až na nějakou zaokrouhlovaíc chybu. To je to, co se tady řešilo s tím `tf.where`.

File: neuralmonkey/attention/scaled_dot_product.py (hunk 4 above)
Comment (jlibovicky) [reply neuralmonkey/attention/scaled_dot_product.py]: Tohle hlavně není dot product, ale weighted average. Poslední dimenze vah se broadcastuje a pak se to sečte.

Jsou tam místa, kde se podobně píše dot product, většina z nich jako dotproduct napsat nejde, protože to o dimenzi nesedí.

Comment (jindrahelcl) [reply neuralmonkey/attention/scaled_dot_product.py]: Já myslim, že @varisd má pravdu. Pokud děláš matmul matic s tvarem [1, N] a [N, M], dostaneš (vodorovnej) vektor [1, M], kde každej prvek na tom řádku bude skalární součin toho sloupcovýho vektoru s korespondujícím řádkem tý druhý matice. Což odpovídá váženýmu součtu čísel z odpovídajícího "kanálu" přes všechny časy.

Jediný co je na @varisd  návrhu špatně je ta mínus jednička, místo který patří podle mě nula.

File: neuralmonkey/attention/scaled_dot_product.py (hunk 3 above)
Comment (jindrahelcl) [reply neuralmonkey/attention/scaled_dot_product.py]: Tim, že ty logity můžou bejt záporný, tak je jedna cesta to napřed zesoftmaxit a pak renormalizovat.

Druhá cesta by byla vymaskovat to mínus nekonečnama:

```python
if self.attention_mask is not None:
  w_exp = tf.exp(weights)
  w_masked = weigths * self.attention_mask
  weights = tf.log(w_masked)
```

třetí cesta je použít `tf.where` ale teď nevim proč by.

Comment (jindrahelcl) [reply neuralmonkey/attention/scaled_dot_product.py]: EDIT první verze kodu byla blbě, teď už je to snad dobre

File: neuralmonkey/attention/scaled_dot_product.py (hunk 4 above)
Comment (varisd) [reply neuralmonkey/attention/scaled_dot_product.py]: Jo, urcite tam ma byt nula.

File: neuralmonkey/attention/scaled_dot_product.py (hunk 3 above)
Comment (varisd) [reply neuralmonkey/attention/scaled_dot_product.py]: Ok, v pohode.

Comment (jindrahelcl) [reply neuralmonkey/attention/scaled_dot_product.py]: tim editem sem myslel ten kod v minulym komentari (zacinajici slovy "Tim, že logity") A myslim si, že jestli je to dobre, tak bychom to meli pouzit - zbavi nas to toho `1e-08` stejně jako by nás toho zbavilo `tf.where` a navíc to bude bez potenciálního nebezpečí při použití ve while cyklu

Comment (jindrahelcl) [reply neuralmonkey/attention/scaled_dot_product.py]: Možná by bylo ještě lepší udělat pomocnou funkci `masked_softmax`, abychom nemuseli ty exponenciely počítat dvakrát za sebou:

```python
def masked_softmax(logits, mask, dim):
  masked = tf.exp(logits) * mask
  return masked / tf.reduce_sum(masked, dim)
```
